File: algorithms/M/Maximum Number of Non-Overlapping Subarrays With Sum Equals Target/Maximum Number of Non-Overlapping Subarrays With Sum Equals Target.py
'''
greedy, prefix sum with hashtable
O(n), O(n)
'''
class Solution:
    def maxNonOverlapping(self, nums: List[int], target: int) -> int:
        # hash set to record previously encountered prefix sums
        prefix_sums = {0}
        
        res = prefix_sum = 0
        for num in nums:
            prefix_sum += num
            if prefix_sum - target in prefix_sums:
                res += 1
                # greedily discard prefix sums before num
                # thus not considering subarrays that start at before num 
                prefix_sums = {prefix_sum} 
            else:
                prefix_sums.add(prefix_sum)
        return res

File: algorithms/M/Maximum Number of Non-Overlapping Subarrays With Sum Equals Target/Maximum Number of Non-Overlapping Subarrays With Sum Equals Target.js
var maxNonOverlapping = function(nums, target) {
    const seen = new Set();
    let total = 0, result = 0;
    
    for(let n of nums) {
        total += n;
        
        if(total === target || seen.has(total - target)) {
            total = 0;
            result++;
            seen.clear()
        } else seen.add(total)
    }
    return result;
};
